stack-graphs/src/cycles.rs

```diff
@@ -129,71 +129,150 @@ where
-
-impl<A: Appendable + Clone> AppendingCycleDetector<A> {
+impl<H> AppendingCycleDetector<H> {
     pub fn new() -> Self {
         Self {
             appendages: List::empty(),
         }
     }
 
-    pub fn from(appendables: &mut Appendables<A>, appendage: A) -> Self {
+    pub fn from(appendables: &mut Appendables<H>, path: PartialPath) -> Self {
+        let h = appendables.interned.add(path);
         let mut result = Self::new();
-        result.appendages.push_front(appendables, appendage);
+        result
+            .appendages
+            .push_front(&mut appendables.elements, InternedPathOrHandle::Owned(h));
         result
     }
 
-    pub fn append(&mut self, appendables: &mut Appendables<A>, appendage: A) {
-        self.appendages.push_front(appendables, appendage);
+    pub fn append(&mut self, appendables: &mut Appendables<H>, appendage: H) {
+        self.appendages.push_front(
+            &mut appendables.elements,
+            InternedPathOrHandle::Database(appendage),
+        );
     }
+}
 
+impl<H> AppendingCycleDetector<H>
+where
+    H: Clone,
+{
     /// Tests if the path is cyclic. Returns a vector indicating the kind of cycles that were found.
     /// If appending or concatenating all fragments succeeds, this function will never raise and error.
-    pub fn is_cyclic(
+    pub fn is_cyclic<'a, A, Db>(
         &self,
         graph: &StackGraph,
         partials: &mut PartialPaths,
-        ctx: &mut A::Ctx,
-        appendables: &mut Appendables<A>,
-    ) -> Result<EnumSet<Cyclicity>, PathResolutionError> {
+        db: &'a Db,
+        appendables: &mut Appendables<H>,
+    ) -> Result<EnumSet<Cyclicity>, PathResolutionError>
+    where
+        A: Appendable + 'a,
+        Db: ToAppendable<H, A>,
+    {
         let mut cycles = EnumSet::new();
 
-        let end_node = match self.appendages.clone().pop_front(appendables) {
-            Some(appendage) => appendage.end_node(ctx),
+        let end_node = match self.appendages.clone().pop_front(&mut appendables.elements) {
+            Some(appendage) => appendage.end_node(db, &appendables.interned),
             None => return Ok(cycles),
         };
 
         let mut maybe_cyclic_path = None;
-        let mut appendages = self.appendages;
+        let mut remaining_appendages = self.appendages;
+        // Unlike the stored appendages, which are stored in a shared arena, we use a _local_
+        // buffer to collect the prefix appendages that we collect for possible cycles. This is
+        // to prevent adding elements to the shared arena for every invocation of this method,
+        // because they would remain in the arena after the method returns. We take care to
+        // minimize (re)allocations by (a) only allocating when a possible cycle is detected,
+        // (b) reserving all necessary space before adding elements, and (c) reusing the buffer
+        // between loop iterations.
+        let mut prefix_appendages = Vec::new();
```

Comment:
Your explanation here makes sense. Is there a minimum lower bound that we can use to preallocate a useful amount of element space for this vector?

Reply:
Good question. Thinking about it, I don't know if that would improve things. The empty vector doesn't allocate at all, I think. The means only when we find a potential cycle and are going to use the buffer, do we allocate, and we'll allocate exactly the right size. So, if there are no possible cycles, we don't allocate at all with the current code. If there's one cycle, it will allocate the right amount. Only if there are multiple possible cycles, and any of the following cycles is strictly larger than the first one, will a reallocation take place. I don't know how often that happens, but if we assume not that often, I think starting with an empty vector is still a good idea here.